plugins/module_utils/s3.py

```diff
@@ -100,3 +106,64 @@ def validate_bucket_name(module, name):
     if name[-1] not in string.ascii_lowercase + string.digits:
         module.fail_json(msg='bucket names must begin and end with a letter or number')
     return True
+
+
+# To get S3 connection, in case of dealing with ceph, dualstack, etc.
+def is_fakes3(endpoint_url):
+    """ Return True if endpoint_url has scheme fakes3:// """
+    result = False
+    if endpoint_url is not None:
+        result = urlparse(endpoint_url).scheme in ('fakes3', 'fakes3s')
+    return result
+
+
+def get_s3_connection(module, aws_connect_kwargs, location, ceph, endpoint_url, sig_4=False):
+    params = dict(
+        conn_type='client',
+        resource='s3',
+        region=location,
+        endpoint=endpoint_url,
+        **aws_connect_kwargs
+    )
+    if ceph:  # TODO - test this
+        ceph = urlparse(endpoint_url)
+        use_ssl = bool(ceph.scheme == 'https')
+        params.update(
+            dict(
+                use_ssl=use_ssl
+            )
+        )
+    elif is_fakes3(endpoint_url):
+        fakes3 = urlparse(endpoint_url)
+        protocol = "http"
+        port = fakes3.port or 80
+        if fakes3.scheme == 'fakes3s':
+            protocol = "https"
+            port = fakes3.port or 443
+        endpoint_url = f"{protocol}://{fakes3.hostname}:{to_text(port)}"
+        use_ssl = bool(fakes3.scheme == 'fakes3s')
+        params.update(
+            dict(
+                endpoint=endpoint_url,
+                use_ssl=use_ssl,
+            )
+        )
```

Comment:
It would probably be good to refactor this piece into its own function which can be cleanly unit tested